### PR TITLE
Idbio 35 opt in screens

### PR DESCRIPTION
**Pull request by metareven**
Created a flow for enrolling into, using and administrating awesome possum.
In this example version, there is a 50% chance of getting correctly identified and 50% change of not being identified.

How to test:
Run the app on your favorite android device
Log in and press the button to enroll
?????
Profit

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -36,10 +36,13 @@
...
                 <category android:name="android.intent.category.LAUNCHER" />
             </intent-filter>
         </activity>
+        <activity android:name=".EnrollActivity" android:theme="@style/Theme.AppCompat.Dialog.Alert"></activity>
+        <activity android:name=".ReadMoreActivity" android:theme="@style/Theme.AppCompat.Dialog.Alert" />
+        <activity android:name=".AgreeToUsePossumActivity" android:theme="@style/Theme.AppCompat.Dialog.Alert" />
+        <activity android:name=".PossumInfoActivity"></activity>
     </application>
 
 </manifest>
```
Comment (jorunfa) [inline app/src/main/AndroidManifest.xml]: Add newline.

File: app/src/main/java/com/telenordigital/protoman/AgreeToUsePossumActivity.kt [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+class AgreeToUsePossumActivity : AppCompatActivity() {
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_agree_to_use_possum)
+
+        val yesButton = findViewById<Button>(R.id.AgreeToUsePossumYesButton)
+        yesButton.setOnClickListener{
+            val possumGather = PossumGather(this,"ProtoMan")
+            if(possumGather.hasMissingPermissions(this)){
```
Comment (jorunfa) [inline app/src/main/java/com/telenordigital/protoman/AgreeToUsePossumActivity.kt]: Follow Kotlin style guide for spaces?
E.g.
`if(possumGather.hasMissingPermissions(this)){` ->
`if (possumGather.hasMissingPermissions(this)) {`

**Comment by jorunfa** [comment]
Review in progres...
Pushed cb46be8 to fix the relogin.

File: app/src/main/java/com/telenordigital/protoman/AgreeToUsePossumActivity.kt [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+import com.telenor.possumgather.PossumGather
+
+class AgreeToUsePossumActivity : AppCompatActivity() {
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_agree_to_use_possum)
+
+        val yesButton = findViewById<Button>(R.id.AgreeToUsePossumYesButton)
+        yesButton.setOnClickListener{
```
Comment (jorunfa) [inline app/src/main/java/com/telenordigital/protoman/AgreeToUsePossumActivity.kt]: `yesButton.setOnClickListener{` ->
`yesButton.setOnClickListener {`

File: app/src/main/java/com/telenordigital/protoman/AgreeToUsePossumActivity.kt [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            val possumGather = PossumGather(this,"ProtoMan")
+            if(possumGather.hasMissingPermissions(this)){
+                possumGather.requestNeededPermissions(this)
+            }else{
+                activate()
+            }
+        }
+
+        val laterButton = findViewById<Button>(R.id.AgreeToUsePossumLaterButton)
+        laterButton.setOnClickListener{
```
Comment (jorunfa) [inline app/src/main/java/com/telenordigital/protoman/AgreeToUsePossumActivity.kt]: `laterButton.setOnClickListener{` ->
`laterButton.setOnClickListener {`

File: app/src/main/java/com/telenordigital/protoman/AgreeToUsePossumActivity.kt [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+            val intent = Intent(this,EnrollActivity::class.java)
```
Comment (jorunfa) [inline app/src/main/java/com/telenordigital/protoman/AgreeToUsePossumActivity.kt]: `val intent = Intent(this,EnrollActivity::class.java)`->
`val intent = Intent(this, EnrollActivity::class.java)`

**Review by jorunfa — COMMENTED** [review]
Not reviewed everything yet, will do a fuller review as soon as I can. Got some feedback already, though.

1. Prototype works in that it demonstrates the intent 💪🎉

1. You do not have to re-authenticate if you're in _possum settings_ (and I guess any other activity than the signed in activity?); is this fine for a prototype or should we look into [ProcessLifecycleOwner](https://developer.android.com/reference/android/arch/lifecycle/ProcessLifecycleOwner)? I am not sure myself (found it at: https://stackoverflow.com/a/19920353/2148380).

2. Nevermind, you fixed it.
I get blue text on grey background - super hard to read
image

3. You really hate spaces it seems 😢. Please consider adding some, all over the place. https://android.github.io/kotlin-guides/style.html

**Comment by metareven** [comment]
> You do not have to re-authenticate if you're in possum settings (and I guess any other activity than the signed in activity?); is this fine for a prototype or should we look into ProcessLifecycleOwner? I am not sure myself (found it at: https://stackoverflow.com/a/19920353/2148380).

Sure I can implement that, I just wanted to do it quick and dirty in order to finish it as fast as possible. We are however ahead of schedule, so I could probably do it correctly the first time around.

**Comment by metareven** [comment]
> You really hate spaces it seems . Please consider adding some, all over the place. https://android.github.io/kotlin-guides/style.html

Hate is a strong word, but I strongly dislike spaces before parantheses, brackets and so on. I will read through the best practices and make changes

File: app/src/main/java/com/telenordigital/protoman/AgreeToUsePossumActivity.kt [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
+        for (result in grantResults) {
+            if (result != PERMISSION_GRANTED) {
+                Toast.makeText(this, "Failed setting permissions. Unable to activate", Toast.LENGTH_SHORT).show()
+                return
+            }
+        }
+        activate()
+
```
Comment (jorunfa) [inline app/src/main/java/com/telenordigital/protoman/AgreeToUsePossumActivity.kt:45]: Suggestion: delete line.

File: app/src/main/java/com/telenordigital/protoman/ReadMoreActivity.kt [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+import android.os.Bundle
+import android.widget.Button
+
+class ReadMoreActivity : AppCompatActivity() {
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_read_more)
+
+        val button = findViewById<Button>(R.id.ReadMorePopupButton)
```
Comment (jorunfa) [inline app/src/main/java/com/telenordigital/protoman/ReadMoreActivity.kt:13]: I'm not sure what I think is better. This or alt:
```
findViewById<Button>(R.id.ReadMorePopupButton).setOnClickListener {
    finish()
}
```
Perhaps no point in saving it to a variable if we only use it once? Just a suggestion, though, I think either is fine.

File: app/src/main/java/com/telenordigital/protoman/SignedInActivity.kt [hunk 8]
```diff
@@ -10,18 +12,42 @@ import com.telenor.connect.id.ConnectTokensStateTracker
...
         }
+        val usesPossum = getSharedPreferences(getString(R.string.preference_id), Context.MODE_PRIVATE).getBoolean(getString(R.string.is_possum_enabled), false)
+        val possumButton = findViewById<Button>(R.id.possum_button)
+        if (!usesPossum) {
+            val intent = Intent(this, EnrollActivity::class.java)
+            possumButton.setText("Start Using Awesome Possum")
+            possumButton.setOnClickListener {
+                startActivity(intent)
+            }
+
```
Comment (jorunfa) [inline app/src/main/java/com/telenordigital/protoman/SignedInActivity.kt:38]: Suggestion: Remove empty line.

File: app/src/main/res/layout/activity_agree_to_use_possum.xml [hunk 9]
```diff
@@ -0,0 +1,60 @@
+<?xml version="1.0" encoding="utf-8"?>
+<android.support.constraint.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    tools:context=".AgreeToUsePossumActivity">
+
+    <Button
+        android:id="@+id/AgreeToUsePossumYesButton"
```
Comment (jorunfa) [inline app/src/main/res/layout/activity_agree_to_use_possum.xml:10]: I'm more used to lower case, underline separated, id names, but doesn't matter for prototype though. Argue if you disagree 🤓📝🤓.
image

https://github.com/umesh0492/android-guidelines/blob/master/project_and_code_guidelines.md

File: app/src/main/res/layout/activity_enroll.xml [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+        android:layout_marginStart="148dp"
+        android:layout_marginEnd="148dp"
+        android:layout_marginBottom="8dp"
+        android:text="YES!"
+        app:layout_constraintBottom_toTopOf="@+id/NoLink"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toBottomOf="@+id/readMoreLink" />
+
+</android.support.constraint.ConstraintLayout>
```
Comment (jorunfa) [inline app/src/main/res/layout/activity_enroll.xml:132]: Missing newline.

File: app/src/main/res/layout/activity_possum_info.xml [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+        android:id="@+id/PossumInfoOptOutButton"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginStart="8dp"
+        android:text="Opt Out"
+        app:layout_constraintBottom_toBottomOf="@+id/PossumInfoOKButton"
+        app:layout_constraintStart_toEndOf="@+id/PossumInfoOKButton"
+        app:layout_constraintTop_toTopOf="@+id/PossumInfoOKButton" />
+
+</android.support.constraint.ConstraintLayout>
```
Comment (jorunfa) [inline app/src/main/res/layout/activity_possum_info.xml:70]: Newline

File: app/src/main/res/layout/activity_read_more.xml [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    <Button
+        android:id="@+id/ReadMorePopupButton"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="16dp"
+        android:text="OK"
+        app:layout_constraintEnd_toEndOf="@+id/ReadMoreText"
+        app:layout_constraintStart_toStartOf="@+id/ReadMoreText"
+        app:layout_constraintTop_toBottomOf="@+id/ReadMoreText" />
+</android.support.constraint.ConstraintLayout>
```
Comment (jorunfa) [inline app/src/main/res/layout/activity_read_more.xml:42]: newline

**Review by jorunfa — APPROVED** [review]
I love arguing, so I added a bunch of small notes.

You can ignore as many of them as you wish, I think this is still fine for a prototype. I think it's achieving it's goal to create something visual and to hold and feel, that will help us as a team discuss and conclude what we're making.

Good job 😘

**Pull request merged**